### ISSUE TITLE
Redesign SolutionTransfer

### ISSUE THREAD

**Issue by marcfehling** [dealii/dealii#15280]
Part of https://github.com/dealii/dealii/issues/15225#issuecomment-1556079986.

Separate issue for the developer workshop.

---

I agree in making the SolutionTransfer class more flexible. Though in a different context, @Rombur and I were discussing approaches starting with https://github.com/dealii/dealii/pull/11132, where we contemplated on how to do a transfer from a FE_Nothing to a different element that does not represent nullspace. I hope we can work on the prototype in the upcoming hackathon.

I would suggest to let SolutionTransfer only provide the external scaffolding for the whole transfer operation. For the rework, I suggest we realise the preparation and recovery of the data purely via callback functions: The user needs to specify how the data will be prepared/recovered by providing a function for each operation themselves. SolutionTransfer then calls these callbacks at the required place. Ideally, these callbacks will be interchangeable between serial and parallel implementations of SolutionTransfer.

We can provide a library of such functions for different purposes: one for preparation and one for recovery of cell-related data (as now with CellDataTransfer), dof-related data (as now with SolutionTransfer), quadrature-related data (as now with ContinuousQuadratureDataTransfer), and with some finite element activation logic (as now with FieldTransfer).

We now have four different classes in the parallel::distributed namespace that kind of do the same thing, and I would like to help to merge them into a more flexible class. This will make the code more easily maintainable, more flexible, and avoids code duplication. This would also lay out the path for your vision, @kronbichler.

The question of course remains if the idea I have in my head will actually work out...

**Comment by bangerth** [comment]
Here's what transpired in my brain during @marcfehling 's presentation:
```
template <int dim, int spacedim>
class MeshAttachedData {
public:
  /**
   * Constructor. The argument is the triangulation class to whose cells we want to
   * attach data.
   */
  MeshAttachedData (const Triangulation<dim,spacedim> &tria);

  /**
   * Add another data set to the data that relates to the triangulation. The two arguments
   * describe ways to serialize and deserialize data that corresponds to each cell.
   */
  void add_data_source (const std::function<std::vector<char> (const Triangulation::cell_iterator &cell)> &pack_cell_data,
                                           const std::function<void (const Triangulation::cell_iterator &cell,
                                                                                         const std::vector<char> &buffer)> &unpack_cell_data);

  /**
   * Add another data set characterized by a DoFHandler and a solution vector.
   */
  template <typename NumberType>
  void add_data_source (const DoFHandler &dof_handler,
                                            const ReadVector<NumberType> &vector)
  {
    ...implement in terms of calling the function above with two lambda functions that wrap the dof_handler and vector...
  }

  ... perhaps more ways to attach sources ...

  /**
   * Add another data set characterized by a particle handler.
   */
  void add_data_source (const ParticleHandler &particle_handler)
  {
    ...implement in terms of calling the first function above with two lambda functions that wrap the particle_handler...
  }

  /**
   * Actually call first of each pair of lambda functions attached above, and store the resulting data somewhere.
   * This may take into account which cells are slated for refinement or coarsening.
   */
  void attach_data ();

  /**
   * Actually call second of each pair of lambda functions attached above, getting the data where it was stored.
   */
  void detach_data ();


  /**
   * Call attach_data() and put the resulting data into a stream (or do the equivalent without storing the data
   * somewhere in the middle -- not important to specify the details).
   */
  void serialize (std::ofstream &);

  /**
   * Get the data from the input stream and then call detach_data() with it (or do the equivalent without storing the data
   * somewhere in the middle -- not important to specify the details).
   */
  void serialize (std::ifstream &);

private:
  std::vector<std::pair<std::function<...>,      // pack cell data callback
                                        std::function<...>>     // unpack cell data callback
                     > data_sources;
};
```

Comments welcome. This is just an idea -- all subject to revision, negotiation, or to just being ignored :-)

**Comment by marcfehling** [comment]
This is the [freeplane](https://docs.freeplane.org/) mindmap from today's discussion (slightly adjusted): [SolutionTransfer.mm.zip](https://github.com/dealii/dealii/files/11931266/SolutionTransfer.mm.zip)

**Comment by tjhei** [comment]
> Comments welcome.

I like it.

Minor comments:
- one function should be called deserialize not serialize
- attach/detach_data() might not be necessary because the class would do the action during/after Triangulation::execute_coarsening_and_refinement() (or maybe we do a interpolate() afterwards?)
- ``add_data_source (DoFHandler &,const ReadVector<NumberType> &vector)`` doesn't work this way because we need to write into the vector after the transfer, right? So we can not use ReadVector or const.

**Comment by bangerth** [comment]
Yes to points 1 & 3. For 2: You want these functions because when you serialize/deserialize, the `Triangulation::execute_coarsening_and_refinement() ` function may not be called.

If the `SolutionTransfer` class was built on this thing here, you'd connect `attach/detach_data()` to the various signals of the `Triangulation` class.

**Comment by kronbichler** [comment]
Regarding the interface:
 ```c++
  template <typename VectorType>
  void add_data_source (const DoFHandler &dof_handler,
                        const VectorType &vector)
```
Given the progress we saw by introduing `ReadVector` and as a follow-up to @tjhei's comment, I guess we should implement all heavy-lifting with an internal function templated on `NumberType`, and only do a final copy/access into the vector in a header file? I really think that we should also have all write access inside deal.II's function without the template on the vector type.

**Comment by marcfehling** [comment]
I like the idea to have one class that deals with all the ways to attach data for transfer/serialization. It is also easily extendible for users, especially when providing lambda functions.

We already have a class `CellAttachedData`, with slightly different responsibilities, which used to be internal to `p::d::T`. I think this name would be also a good pick, since we need to provide a mechanism how to prepare data *on each cell independently*.

**Comment by bangerth** [comment]
Regarding @kronbichler comment:
```
 template <typename VectorType>
 void add_data_source (const DoFHandler &dof_handler,
                       const VectorType &vector)
```
This function would be implemented in the .h file and simply call the other function with two appropriately built lambda functions. The `VectorType` type does not actually leak into the `.cc` files and does not cause instantiations of anything other than this one function itself.

**Comment by bangerth** [comment]
On 7/3/23 08:31, Marc Fehling wrote:
> We already have a class |CellAttachedData|, with slightly different 
> responsibilities, which used to be internal to |p::d::T|. I think this name 
> would be also a good pick, since we need to provide a mechanism how to prepare 
> data /on each cell independently/.

Yes, all names subject to change of course!

**Comment by pcafrica** [comment]
For reference:

_Originally posted by @marcfehling in https://github.com/dealii/dealii/pull/15706#discussion_r1715924502_

**Comment by marcfehling** [comment]
`p4est` is capable to perform the data transfer independent of mesh partitioning with the `p4est_transfer_*` functions. They have been introduced with version 2.0 and their use has been described in section 5.2 of the [2020 paper](https://dl.acm.org/doi/10.1145/3401990) ([preprint](https://arxiv.org/abs/1803.08432)). It just requires a deep copy of the element counts of the before state.

I would suggest to separate the data transfer part from the `execute_coarsening_and_refinement` and `repartition` functions in the redesign. This would make our whole SolutionTransfer apparatus much more flexible and would allow for use cases that @winnifried described on the [mailing list](https://groups.google.com/g/dealii/c/ZgTHuk1b1VU) (both in serial and parallel after #15706).

I am confident that we can also come up with a clever way to implement this change without breaking user code. For now, let's keep this feature in mind.

**Comment by bangerth** [comment]
Worth mentioning: When you called `prepare_for_pure_refinement()`, you didn't pass any arguments to that function; later on, you'd call `interpolate(solution)` and it would take an old solution vector, store information, resize the vector, and store the interpolated solution on the new mesh into the vector. In other words, what `prepare_for_pure_refinement()` did was not defer transfering data, but *storing an operator* that facilitated the later `interpolate()` function. I believe that this operator could be applied multiple times.

I don't know the p4est 2.0 interface, but it seems difficult to imagine how exactly you'd create such a transfer operator in parallel.
